mediatek: add support for Confiabits MT7981

LuisMitaHL opened the pull request:
Confiabits MT7981 is a Wi-Fi 6 router based on MediaTek MT7981.

Specification:
 - SoC: MediaTek MT7981B
 - CPU: 2x 1.3 GHz Cortex-A53
 - Flash: 128 MiB SPI NAND
 - RAM: 256 MiB
 - WLAN: 2.4 GHz, 5 GHz (MediaTek MT7976CN, 802.11ax)
 - Ethernet: 4x 10/100/1000 Mbps MT7531AE (3xLAN, 1xWAN)
 - USB 2.0 port
 - Buttons: 1 Reset button, 1 Mesh button.
 - LEDs: 7x light-blue, 2x warm-white
 - Serial console: internal 4-pin header, 115200 8n1
 - Power: 12 VDC, 1.5 A

Installation:
The stock firmware is OpenWrt-based. If you can reach LuCI or SSH, just use the sysupgrade image. There is a factory image you can use with u-boot.

Confiabits MT7981 is a Bolivian made router and the stock firmware is OpenWrt 21.02 plus Mediatek proprietary drivers.

This is my first time adding OpenWrt mainline support to a router. I've reviewed similar commits and made sure everything is functional.

* UART
* Ethernet WAN
* Ethernet LAN1/2/3
* Wi-Fi 2.4GHz
* Wi-Fi 5GHz
* USB Power
* USB Data
* LEDs
* SPI Flash

enmaskarado commented:
Hello,
I updated the PR to add persistent MAC addresses. I didn't recognise it the first time as it was a bug in the initial version of the stock firmware.

Also I have the same commit for version 23.05, should I prepare another PR? Or do I wait to resolve this first one?

alphaonex86 commented:
Tested-by: BRULE Herman <contact@herman-brule.com>

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -204,6 +204,25 @@ define Device/cetron_ct3003
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-ubi | check-size $$$$(IMAGE_SIZE)
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb3 kmod-mt7981-firmware
```
`kmod-usb2`?  `kmod-usb3` is not enough?

enmaskarado replied:
Good point, I check and if it works I remove the kmod-usb2 package.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,279 @@
+};
+
+&wifi {
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
Are your wireless MACs in OpenWrt equal to the MACs in stock firmware?

enmaskarado replied:
Yes

csharper2005 commented:
> Tested-by: BRULE Herman [contact@herman-brule.com](mailto:contact@herman-brule.com)

What is the system `load average` on your device? You can check it with `uptime` command.

alphaonex86 commented:
> > Tested-by: BRULE Herman [contact@herman-brule.com](mailto:contact@herman-brule.com)
> 
> What is the system `load average` on your device? You can check it with `uptime` command.

Constant load average of 1.0

enmaskarado commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,279 @@
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				macaddr: macaddr@a {
```
The stock partition comes that way. I have tested it on several routers and confirmed with hexdump the position.

csharper2005 replied:
1. Ok, your MAC is at 0x4. And not at the position 0xa. Why the label is `macaddr@a`? And not `macaddr@4`. IOW, what means `@a`? Of course, this does not affect anything other than the general logic and understandability of the code.
2. Unusually, 0x4 is a WLAN 2.4 MAC.  Do I understand correctly that WLAN 2.4 MAC = LAN MAC in the stock firmware? We would have fewer questions with a MAC table in the commit message.

enmaskarado replied:
I thought the MAC was at 0xa all this time, but it's actually at 0x4. I've corrected it, thanks
Regarding the MACs, all MACs are the same as in the stock firmware.

enmaskarado commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -204,6 +204,25 @@ define Device/cetron_ct3003
+  DEVICE_DTS := mt7981b-confiabits-mt7981
+  DEVICE_DTS_DIR := ../dts
+  IMAGES := sysupgrade.bin
+  SUPPORTED_DEVICES += confiabits,mt7981
```
No, I've only checked other commits of similar devices and those commits have those lines. I don't know if any of those lines are not essential.

enmaskarado commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -204,6 +204,25 @@ define Device/cetron_ct3003
+  DEVICE_MODEL := MT7981
+  DEVICE_DTS := mt7981b-confiabits-mt7981
+  DEVICE_DTS_DIR := ../dts
+  IMAGES := sysupgrade.bin
```
Ok

enmaskarado commented:
The router also suffers from the load-average bug seen at PR #13703 
echo 0 > /sys/bus/usb/devices/usb2/authorized fixes the bug.

As the PR creates a new file in preinit, I didn't think it would be convenient to create another one as long as that PR has not been accepted. I had a plan to prepare another PR to add this router to 07_fix_usb, but if there is a better plan I will gladly follow it.

csharper2005 commented:
> The router also suffers from the load-average bug seen at PR #13703 echo 0 > /sys/bus/usb/devices/usb2/authorized fixes the bug.
> 

@Linaro1985 suggested a better solution (adding a `mediatek,u3p-dis-msk = <0x1>;` to the `&xhci` section of dts). You can try this.
```
&xhci {
  status = "okay";
  mediatek,u3p-dis-msk = <0x1>;
};
```

enmaskarado commented:
Thank you, I will try it.

enmaskarado commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,279 @@
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 1 GPIO_ACTIVE_LOW>;
```
Of course, I hadn't noticed the pattern.

enmaskarado commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,279 @@
+
+			partition@380000 {
+				label = "FIP";
+				reg = <0x380000 0x0200000>;
```
No, I will set it to `ro`

enmaskarado commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,279 @@
+			label = "lan1";
+
+			nvmem-cell-names = "mac-address";
+			nvmem-cells = <&macaddr>;
```
I removed the lines.

enmaskarado commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,279 @@
+
+			partition@100000 {
+				label = "u-boot-env";
+				reg = <0x0100000 0x0080000>;
```
I configured the partition as `ro`.

enmaskarado commented:
The fix on the DTS worked very well, thank you very much. i have updated the PR.

enmaskarado commented:
As a note, I have removed the factory image and added `mediatek,mt7981-spim-snand-2500wan-gmac2-rfb` as a supported device.

The factory image has a confusing name and using that image on the stock firmware will brick the router (the sysupgrade image is sufficient). And the supported device name is because the stock firmware uses that one.

starsaminf commented:
Tested-by: Samuel Loza <samuel.loza26@gmail.com>

```
BusyBox v1.36.1 (2023-10-09 21:45:35 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt 23.05.0, r23497-6637af95aa
 -----------------------------------------------------
root@OpenWrt:~# ubus call system board
{
	"kernel": "5.15.134",
	"hostname": "OpenWrt",
	"system": "ARMv8 Processor rev 4",
	"model": "Confiabits MT7981",
	"board_name": "confiabits,mt7981",
	"rootfs_type": "squashfs",
	"release": {
		"distribution": "OpenWrt",
		"version": "23.05.0",
		"revision": "r23497-6637af95aa",
		"target": "mediatek/filogic",
		"description": "OpenWrt 23.05.0 r23497-6637af95aa"
	}
}
root@OpenWrt:~# uptime 
 20:25:39 up  1:50,  load average: 0.06, 0.03, 0.00
```

csharper2005 commented:
@enmaskarado thanks for your updates.

`Some checks were not successful`. Something went wrong during the checking process. Probably you need to make one more forcepush to restart it.

enmaskarado commented:
Rebased PR.

raenye commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -124,7 +142,7 @@
 	#size-cells = <0>;
 
-	switch: switch@0 {
+	switch: switch@31 {
```
Sorry for nagging again. Addresses are supposed to be in hex.
```suggestion
	switch: switch@1f {
```

enmaskarado replied:
Done, thanks

enmaskarado commented:
Hello,
Thanks for the recommendations. I'm going to rebase and squash all in one commit. How can I give you credits in the commit message?

raenye commented:
> I'm going to rebase and squash all in one commit.

👍 

> How can I give you credits in the commit message?

Thanks, but no need. You did the work.

LuisMitaHL committed:
mediatek: add support for Confiabits MT7981

Confiabits MT7981 is a Wi-Fi 6 router based on MediaTek MT7981.

Specification:
 - SoC: MediaTek MT7981B
 - CPU: 2x 1.3 GHz Cortex-A53
 - Flash: 128 MiB SPI NAND
 - RAM: 256 MiB
 - WLAN: 2.4 GHz, 5 GHz (MediaTek MT7976CN, 802.11ax)
 - Ethernet: 4x 10/100/1000 Mbps MT7531AE (3xLAN, 1xWAN)
 - USB 2.0 port
 - Buttons: 1 Reset button, 1 Mesh button.
 - LEDs: 7x light-blue, 2x warm-white
 - Serial console: internal 4-pin header, 115200 8n1
 - Power: 12 VDC, 1.5 A

MAC addresses in stock firmware and in this commit:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| WAN     | 00:0c:43:xx:xx:e1 | label+1   |
| LAN     | 00:0c:43:xx:xx:e0 | label     |
| WLAN 2g | 00:0c:43:xx:xx:e0 | label     |
| WLAN 5g | 02:0c:43:xx:xx:e0 |           |
+---------+-------------------+-----------+
The label MAC was found in 'Factory', 0x4

Installation:
The stock firmware is OpenWrt-based. If you can reach LuCI or SSH, just use the sysupgrade image
with the 'Keep settings' option turned off.

Signed-off-by: Luis Mita <luis@luismita.com>

csharper2005 commented:
@enmaskarado your MAC addresses look like MediaTek default addresses. Have you already checked the MACs on other Confiabits MT7981 device copies? Are they different?

enmaskarado commented:
> @enmaskarado your MAC addresses look like MediaTek default addresses. Have you already checked the MACs on other Confiabits MT7981 device copies? Are they different?

My personal router is from a batch that I believe is one of the earliest. For an event, I have run the firmware on other routers from a new batch, and those come with the new MAC prefix e4:38:44. The MAC of the PR is the same as with the factory firmware.

hauke commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,297 @@
+			label = "wan";
+
+			nvmem-cell-names = "mac-address";
+			nvmem-cells = <&factory_macaddr_4 1>;
```
Does this compile?
`factory_macaddr_4`is not defined in the device tree.
```suggestion
			nvmem-cells = <&macaddr_factory_4 1>;
```

raenye replied:
(that's actually my bad, when I suggested how to eliminate `mac-address-increment`)

enmaskarado replied:
Fixed, thanks

hauke commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,297 @@
+		phy-mode = "2500base-x";
+
+		nvmem-cell-names = "mac-address";
+		nvmem-cells = <&factory_macaddr_4 0>;
```
Does this compile?
`factory_macaddr_4`is not defined in the device tree.
```suggestion
		nvmem-cells = <&macaddr_factory_4 0>;
```

enmaskarado replied:
Fixed

hauke commented:
When you close threads and do not do the change, please comment in them why you haven't dine the change.

csharper2005 commented:
Rebase is required to resolve the merging conflicts.

enmaskarado commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -6,6 +6,12 @@ board=$(board_name)
+confiabits,mt7981)
+	ucidef_set_led_netdev "lan1" "lan1" "blue:lan1" "lan1"
+	ucidef_set_led_netdev "lan2" "lan2" "blue:lan2" "lan2"
+	ucidef_set_led_netdev "lan3" "lan3" "blue:lan3" "lan3"
```
Done, all lan lines contain now `"link tx rx"`

raenye commented on target/linux/mediatek/dts/mt7981b-confiabits-mt7981.dts:
```diff
@@ -0,0 +1,297 @@
+			gpios = <&pio 35 GPIO_ACTIVE_LOW>;
+		};
+	};
+
```
```suggestion
```
Blank line can be removed.

enmaskarado replied:
Fixed

enmaskarado commented:
Suggestion applied

rmilecki commented:
@enmaskarado: we really should have `SPDX-License-Identifier` in `.dts` file. Ideally it should be:
```
// SPDX-License-Identifier: GPL-2.0-only OR MIT
```
@enmaskarado: can you please submit a pull request adding a proper licensing header?
